pkg/main.go

```diff
@@ -1,23 +1,49 @@
 package main
 
 import (
-	"github.com/sarulabs/dingo"
-	"github.com/softonic/homing-pigeon/pkg/generatedServices/dic"
+	"github.com/softonic/homing-pigeon/pkg/messages"
+	"github.com/softonic/homing-pigeon/pkg/middleware"
+	"github.com/softonic/homing-pigeon/pkg/readers"
+	"github.com/softonic/homing-pigeon/pkg/writers"
 	"k8s.io/klog"
+	"os"
+	"strconv"
 )
 
 func main() {
 	klog.InitFlags(nil)
 
-	container, err := dic.NewContainer(dingo.App)
+	bufLen := GetBufferLength("MESSAGE_BUFFER_LENGTH")
+	msgCh1 := make(chan messages.Message, bufLen)
+	msgCh2 := make(chan messages.Message, bufLen)
+
+	bufLen = GetBufferLength("ACK_BUFFER_LENGTH")
+	ackCh1 := make(chan messages.Ack, bufLen)
+	ackCh2 := make(chan messages.Ack, bufLen)
+
+	reader, err := readers.NewReader(msgCh1, ackCh2)
+	if err != nil {
+		panic(err)
+	}
+
+	writer, err := writers.NewWriter(msgCh2, ackCh1)
 	if err != nil {
 		panic(err)
 	}
-	reader := container.GetReader()
-	middleware := container.GetMiddleware()
-	writer := container.GetWriter()
+
+	incomingMiddleware := middleware.NewMiddlewareManager(msgCh1, msgCh2, ackCh1, ackCh2)
 
 	go reader.Start()
-	go middleware.Start()
+	go incomingMiddleware.HandleInput()
```

Comment:
I understand the idea of making middleware manager agnostic to input/output communication. But I didn't find a way to make it more generic, since the channels are typed. The communication between Reader and Writer is through `Message` type channels, while the communication Writer -> Reader is through `Ack` type channels.
The name `incomingMiddleware` here is wrong, it might be `middlewareManager`.

Reply:
I think you can use a generic type like `interface{}`

Reply:
Do we have some news about this?

Reply:
No, we don't. The task TRAF-1354 was pushed back to the backlog due to new priorities and not enough resources.
This subtask of adding middleware between writer and reader was paused, as does not bring value by itself.